### PR TITLE
Setup The Project

### PR DESCRIPTION
**Pull request by Osama-you**
- Add VS Code settings 
- Setup ESlint Configurations  For Server & React 
- Install modules for The server and React  #4
- Run the server on port 5000 
- Setup the React app 
- Setup Travis CI Config

### CODE REVIEW

File: server/utils/jwt.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+const { SECRET_TOKEN } = process.env;
+
+const sign = (data) =>
+  new Promise((resolve, reject) => {
+    jwt.sign(data, SECRET_TOKEN, (err, decoded) => {
+      if (err) return reject(err);
+      return resolve(decoded);
+    });
+  });
```
Comment (Mu7ammadAbed) [inline server/utils/jwt.js:11]: It's common to call it `payload` and it's `encoded`
```suggestion
const sign = (payload) =>
  new Promise((resolve, reject) => {
    jwt.sign(payload, SECRET_TOKEN, (err, encoded) => {
      if (err) return reject(err);
      else return resolve(encoded);
    });
  });
```

File: server/utils/jwt.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+const verify = (token) =>
+  new Promise((resolve, reject) => {
+    jwt.verify(token, SECRET_TOKEN, (err, decoded) => {
+      if (err) return reject(err);
+      return resolve(decoded);
```
Comment (Mu7ammadAbed) [inline server/utils/jwt.js:17]:
```suggestion
      else return resolve(decoded);
```

File: client/src/index.js [hunk 3]
```diff
@@ -1,17 +1,16 @@
...
+
 import App from './App';
-import reportWebVitals from './reportWebVitals';
 
 ReactDOM.render(
-  <React.StrictMode>
-    <App />
-  </React.StrictMode>,
+  <Router>
+    <ConfigProvider locale={frFR}>
```
Comment (Mu7ammadAbed) [inline client/src/index.js:11]: Our website is an RTL Arabic website not french, so we should pass the right direction and language, and even if you imported the right language `'antd/lib/locale/ar_EG'` you should declare it with the appropriate name such as `Arabic`

```suggestion
    <ConfigProvider direction="rtl" locale={Arabic}>
```

File: client/src/index.js [hunk 4]
```diff
@@ -1,17 +1,16 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
-import './index.css';
+import { BrowserRouter as Router } from 'react-router-dom';
+import { ConfigProvider } from 'antd';
+import frFR from 'antd/lib/locale/ar_EG';
```
Comment (Mu7ammadAbed) [inline client/src/index.js:5]:
```suggestion
import Arabic from 'antd/lib/locale/ar_EG';
```

File: client/src/index.js [hunk 5]
```diff
@@ -1,17 +1,16 @@
...
+
 import App from './App';
-import reportWebVitals from './reportWebVitals';
 
 ReactDOM.render(
-  <React.StrictMode>
-    <App />
-  </React.StrictMode>,
+  <Router>
```
Comment (Mu7ammadAbed) [inline client/src/index.js:10]: There's no need to use the `Router` here, call it in the `App.js` file

File: client/public/index.html [hunk 6]
```diff
@@ -1,43 +1,18 @@
...
-    <noscript>You need to enable JavaScript to run this app.</noscript>
-    <div id="root"></div>
-    <!--
-      This HTML file is a template.
-      If you open it directly in the browser, you will see an empty page.
+<head>
+  <meta charset="utf-8" />
+  <!-- <link rel="icon" href="%PUBLIC_URL%/favicon.ico" /> -->
+  <meta name="viewport" content="width=device-width, initial-scale=1" />
+  <meta name="theme-color" content="#000000" />
```
Comment (Mu7ammadAbed) [inline client/public/index.html:8]: Write the right description here
```suggestion
    <meta name="description" content="Rawdaty is a ...." />

```

File: client/public/index.html [hunk 7]
```diff
@@ -1,43 +1,18 @@
...
-  </head>
-  <body>
-    <noscript>You need to enable JavaScript to run this app.</noscript>
-    <div id="root"></div>
-    <!--
-      This HTML file is a template.
-      If you open it directly in the browser, you will see an empty page.
+<head>
+  <meta charset="utf-8" />
+  <!-- <link rel="icon" href="%PUBLIC_URL%/favicon.ico" /> -->
```
Comment (Mu7ammadAbed) [inline client/public/index.html:6]: remove this line for now

File: .travis.yml [hunk 8]
```diff
@@ -0,0 +1,15 @@
+
+language: node_js
+node_js:
+  - lts/*
+services:
+  - postgresql
+before_script:
+  - psql -c "CREATE DATABASE rawdaty;" -U postgres
+  - psql -c "CREATE USER rawdatyuser WITH PASSWORD '123456';" -U postgres
```
Comment (Mu7ammadAbed) [inline .travis.yml:9]:
```suggestion
  - psql -c "CREATE USER rawdatyuser WITH PASSWORD '123456';" -U postgres 
  - "npm run build:db"

```

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (Mu7ammadAbed) [inline .travis.yml:1]: Remove this line

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+  - postgresql
+before_script:
+  - psql -c "CREATE DATABASE rawdaty;" -U postgres
+  - psql -c "CREATE USER rawdatyuser WITH PASSWORD '123456';" -U postgres 
+env: 
+  global: 
+  - PGPORT=5432
+test: 
+  adapter: postgresql
+  database: rawdaty
```
Comment (Mu7ammadAbed) [inline .travis.yml:15]: Add a new line after this line.

**Review by Mu7ammadAbed — CHANGES_REQUESTED** [review]
Great job everyone, but unfortunately there are some tiny mistakes that need to review, and in addition to every comment I mentioned there are some files that need to add a new line at the end of the file.

File: client/.eslintrc.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    "ecmaFeatures": {
+      "jsx": true
+    },
+    "ecmaVersion": 12,
+    "sourceType": "module"
+  },
+  "plugins": ["react", "prettier"],
+  "rules": {
+    //  allow .js extensions for JSX
+    "react/state-in-constructor": [0, "always"],
```
Comment (Mu7ammadAbed) [inline client/.eslintrc.json:20]: Since we will use hooks not classes I think there's no need for this rule

File: client/src/App.js [hunk 12]
```diff
@@ -1,25 +1,6 @@
...
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
-    </div>
-  );
-}
+const App = () => <div className="App">روضتي هنا</div>;
```
Comment (Mu7ammadAbed) [inline client/src/App.js:4]: import the Router and wrap the App component with it

Comment (Mu7ammadAbed) [reply client/src/App.js:4]: `import { BrowserRouter as Router } from 'react-router-dom';`

File: .eslintrc.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    "airbnb-base",
+    "prettier"
+  ],
+  "parserOptions": {
+    "ecmaVersion": 12
+  },
+  "plugins": [
+    "prettier"
+  ],
+  "rules": {
```
Comment (Israa91) [inline .eslintrc.json:12]: Why you don't put this rule here :
```
   "react/jsx-filename-extension":[
            1,
            {"extensions":[".js",".jsx"]}
        ],
```

File: server/app.js [hunk 14]
```diff
@@ -19,6 +20,14 @@ app.use(express.urlencoded({ extended: false }));
 
 app.use('/api/v1/', routes);
 
+app.use(express.static(join(__dirname, '..', 'client', 'build')));
+
+if (process.env.NODE_ENV === 'production') {
+  app.get('*', (req, res) => {
+    res.sendFile(join(__dirname, '..', 'client', 'build', 'index.html'));
```
Comment (Israa91) [inline server/app.js:26]: There is no build folder in client folder , Does you mean bublic instead of build ?

File: .eslintrc.json (hunk 13 above)
Comment (Israa91) [reply .eslintrc.json:12]: Oh , its ok I found it in client side . good job

File: server/app.js (hunk 14 above)
Comment (Mu7ammadAbed) [reply server/app.js:26]: Because this folder will be created when you build your app and in `production` we run the `build` script

**Comment by Israa91** [comment]
Good job my colleagues , I hope in the next time you put a clear description about the important things you do , especially new things that we didn't use it before .

File: server/app.js (hunk 14 above)
Comment (Mu7ammadAbed) [reply server/app.js:26]: Screenshot from 2021-02-02 11-00-49

Comment (Osama-you) [reply server/app.js:26]: When we are in the production stage we should build the react app, and that will create a build file.

**Review by alaa-sahloub — APPROVED** [review]

**Review by Mu7ammadAbed — APPROVED** [review]
Good job Osama and everyone but next time and especially when you push a big `PR` like this you MUST write a better and more descriptive description and insert pics and code snippets for what you did.

**Review by Israa91 — APPROVED** [review]

**Pull request merged**